### PR TITLE
New Terraform tutorial for Anthos Service Mesh

### PR DESCRIPTION
**Pull request by cody-clark**
Adds tutorial for installing Anthos Service Mesh (in-cluster control plane) with Terraform

### CODE REVIEW

**Comment by ToddKopriva** [comment]
@cody-clark , the content of the document says that @ameer00 is the author. Are you a co-author? Should I be working with Ameer or you on the review?

**Comment by cody-clark** [comment]
Hi, @ToddKopriva , co-author is a little strong since the draft is fully @ameer00 's, but I'm happy to work through the review process with you. There's actually two more Terraform tutorials incoming that have a good bit of overlap, so I can make take any changes here back to those drafts, too.

**Comment by cody-clark** [comment]
Hi, @ToddKopriva , is there anything I need to do to help move this along? Thanks!

**Comment by ToddKopriva** [comment]
> Hi, @ToddKopriva , is there anything I need to do to help move this along? Thanks!

No. This is in the queue. I'll review it soon.

File: tutorials/terraform-asm-in-cluster.md [hunk 1]
```diff
@@ -0,0 +1,445 @@
...
+          network_policy          = false
+          identity_namespace      = "enabled"
+          cluster_resource_labels = { "mesh_id" : "proj-${data.google_project.project.number}" }
+          node_pools = [
+            {
+              name         = "asm-node-pool"
+              autoscaling  = false
+              auto_upgrade = true
+              # ASM requires minimum 4 nodes and e2-standard-4
+              node_count   = 2
```
Comment (ToddKopriva) [inline tutorials/terraform-asm-in-cluster.md:243]: The comment seems to contradict the node_count setting.

This seeming discrepancy exists in the other documents, too.

File: tutorials/terraform-asm-in-cluster.md [hunk 2]
```diff
@@ -0,0 +1,445 @@
...
+        kubectl --context=${CLUSTER_1} -n online-boutique wait --for=condition=available --timeout=5m deployment paymentservice
+        kubectl --context=${CLUSTER_1} -n online-boutique wait --for=condition=available --timeout=5m deployment productcatalogservice
+        kubectl --context=${CLUSTER_1} -n online-boutique wait --for=condition=available --timeout=5m deployment shippingservice
+        kubectl --context=${CLUSTER_1} -n online-boutique wait --for=condition=available --timeout=5m deployment cartservice
+        kubectl --context=${CLUSTER_1} -n online-boutique wait --for=condition=available --timeout=5m deployment loadgenerator
+        kubectl --context=${CLUSTER_1} -n online-boutique wait --for=condition=available --timeout=5m deployment recommendationservice
+
+## Access the Online Boutique app
+
+Access the app using the `istio-ingressgateway` Service hostname:
```
Comment (ToddKopriva) [inline tutorials/terraform-asm-in-cluster.md:409]: The same step in the next document in the series uses this wording:
"Run the following command to get the IP address of the external load balancer:"

File: tutorials/terraform-asm-upgrade.md [hunk 3]
```diff
@@ -0,0 +1,553 @@
...
+        kubectl --context=${CLUSTER_1_CTX} -n online-boutique rollout status -w --timeout=5m deployment emailservice
+        kubectl --context=${CLUSTER_1_CTX} -n online-boutique rollout status -w --timeout=5m deployment frontend
+        kubectl --context=${CLUSTER_1_CTX} -n online-boutique rollout status -w --timeout=5m deployment paymentservice
+        kubectl --context=${CLUSTER_1_CTX} -n online-boutique rollout status -w --timeout=5m deployment productcatalogservice
+        kubectl --context=${CLUSTER_1_CTX} -n online-boutique rollout status -w --timeout=5m deployment shippingservice
+        kubectl --context=${CLUSTER_1_CTX} -n online-boutique rollout status -w --timeout=5m deployment cartservice
+        kubectl --context=${CLUSTER_1_CTX} -n online-boutique rollout status -w --timeout=5m deployment loadgenerator
+        kubectl --context=${CLUSTER_1_CTX} -n online-boutique rollout status -w --timeout=5m deployment recommendationservice
+
+    The output is similar to the following:
```
Comment (ToddKopriva) [inline tutorials/terraform-asm-upgrade.md:431]: This output section is useful. Include it in the other two documents?

**Review by ToddKopriva — COMMENTED** [review]
I left a few comments, most of which are just minor suggestions.

After you've gone through these comments, I'll do a final proofreading pass and then merge the pull request.

File: tutorials/terraform-asm-upgrade.md (hunk 3 above)
Comment (cody-clark) [reply tutorials/terraform-asm-upgrade.md:431]: Added output to the other docs

File: tutorials/terraform-asm-in-cluster.md (hunk 2 above)
Comment (cody-clark) [reply tutorials/terraform-asm-in-cluster.md]: Updated docs to use consistent language

File: tutorials/terraform-asm-in-cluster.md (hunk 1 above)
Comment (cody-clark) [reply tutorials/terraform-asm-in-cluster.md]: Updated all docs to node_count = 4

**Comment by cody-clark** [comment]
Thanks, Todd! I just addressed your comments

**Comment by ToddKopriva** [comment]
> Thanks, Todd! I just addressed your comments

Excellent. I'll do a final proofread now.

**Review by ToddKopriva — APPROVED** [review]

**Comment by ToddKopriva** [comment]
published:
https://cloud.google.com/community/tutorials/terraform-asm-mcp
https://cloud.google.com/community/tutorials/terraform-asm-upgrade
https://cloud.google.com/community/tutorials/terraform-asm-in-cluster

**Pull request merged**